### ISSUE TITLE
wxMemoryFSHandler pictures

### ISSUE THREAD

**Issue by Natulux** [imReker/wxWebViewBlink#3]
Hey,
I was trying to use images in the html file, shown by the webview. Those images come from the memory, using wxMemoryFSHandler.
From the webview sample:

		// Create the memory files
		wxImage::AddHandler(new wxPNGHandler);
		wxMemoryFSHandler::AddFile("logo.png",
			wxBitmap(wxlogo_xpm), wxBITMAP_TYPE_PNG);

		wxMemoryFSHandler::AddFile("page1.htm",
		    "<html><head><title>File System Example</title>"
		    "<link rel='stylesheet' type='text/css' href='memory:test.css'>"
		    "</head><body><h1>Page 1</h1>"
		    "<p><img src='memory:logo.png'></p>"
		    "<p>Some text about <a href='folder/page2.htm'>Page 2</a>.</p></body>");

The img src is in this case with the "memory:" tag. If used like this, it works on webview trident and webview miniBlink.

But if I remove the "memory:" tag from the source path, webview trident is still able to load that image, as long as the starting index.html page had that "memory:" tag. webview miniBlink on the other hand does not load these images, if the tag is missing.

My point is: I cant run through all of html/css/js scripts to change and add "memory:" to each path.
Any idea, where the miniBlink backend differs from the trident one, regarding the loading of relative paths inside the html?
Or any other ideas, how I could come around this?

Thanks.
Natu

**Comment by imReker** [comment]
wxWebViewMiniBlink parse the URL similar to wxWebView in Linux and MacOS, and wxWidgets use different implemention on MSW.
[Here](https://github.com/imReker/wxWebViewBlink/blob/master/WebViewMiniBlink.cpp#L162), wxWebViewMiniBlink will try `Handler::GetName` and URL to find a correct `Handler` then call `getFile`.

But the IE implemention will try `getFile` on all `Handler`s and return the first file it found, see [here](https://github.com/wxWidgets/wxWidgets/blob/master/src/msw/webview_ie.cpp#L1523).

You can skip the `GetName` check and use `getFile` directly like IE implemention.

**Comment by Natulux** [comment]
Hey, thanks for your answer. I think you are hinting me in the right direction, but that was not all.

I now only have one handler added and like you suggested take the handler without checking the name. 

On my test html file I show a picture twice, one with the "memory:" tag and the other without. While the tagged picture runs through the "wxWebViewMiniBlink::OnNavigationInternal" the other does not.
In other words, it isnt event attempting to get a handler. 

Clicking a href link without the "memory:" tag results in wxWebViewMiniBlink:
08:50:11: Navigation request to 'page2.html' (target='')
08:50:11: Running through: wxWebViewMiniBlink::OnOtherLoad
08:50:11: Running through: wxWebViewMiniBlink::OnOtherLoad
08:50:11: Navigation complete; url='about:blank'

And in IE implementation:
08:51:30: Navigation request to 'memory:page2.html' (target='')
08:51:30: Navigation complete; url='memory:page2.html'
08:51:30: Title changed; title='memory:page2.html'
08:51:30: Document loaded; url='memory:page2.html'

So, wxWebViewIE::onActiveXEvent(wxActiveXEvent& evt) does return the path with "memory:" (error or feature?) while wxWebViewMiniBlink::OnNavigationInternal returns the path as is.
Just altering the url has no effect though:

wxString sUrl = url;
if (!sUrl.StartsWith("memory:"))
{
	sUrl.Prepend("memory:");
}
wxWebViewEvent event(wxEVT_WEBVIEW_NAVIGATING, GetId(), sUrl, ConvertFrameId(m_webview, frame));

**Comment by imReker** [comment]
Is URL without 'memory' schema trigger `OnLoadUrlBegin`?

**Comment by Natulux** [comment]
No, it doesn't.
One URL without memory:

09:18:55: Warning: wxWebViewMiniBlink::OnNavigationInternal
09:18:55: Navigation request to 'memory:page2.html' (target='')
09:18:55: Warning: wxWebViewMiniBlink::OnOtherLoad
09:18:55: Warning: wxWebViewMiniBlink::OnOtherLoad
09:18:55: Navigation complete; url='about:blank'
09:18:55: Warning: wxWebViewMiniBlink::OnOtherLoad
09:18:55: Document loaded; url='about:blank'

Another URL with memory:
09:19:57: Warning: wxWebViewMiniBlink::OnNavigationInternal
09:19:57: Navigation request to 'memory:page1.html' (target='')
09:19:57: Warning: wxWebViewMiniBlink::OnOtherLoad
09:19:57: Warning: OnLoadUrlBegin
09:19:57: Warning: running handlers
09:19:57: Warning: handler found: memory:page1.html
09:19:57: Warning: wxWebViewMiniBlink::OnOtherLoad
09:19:57: Navigation complete; url='memory:page1.html'
09:19:57: Warning: wxWebViewMiniBlink::OnOtherLoad
09:19:57: Document loaded; url='memory:page1.html'

**Comment by Natulux** [comment]
To clarify: In my last post, the URL came from href links.
If I type in the url by hand, it is auto completed with the memory tag, which seems to be one effect of forcing it to use the memory handler. :-)

EDIT:
Ah no it wasn't. That came from my changing of the event handler function right on top and has nothing to do with the backend:

```
void WebFrame::OnUrl(wxCommandEvent& WXUNUSED(evt))
wxString sUrl = m_url->GetValue();
	if (!sUrl.StartsWith("memory:"))
	{
		sUrl.Prepend("memory:");
		wxLogWarning("wxWebViewMiniBlink::DoSetPage, Prepended memory tag: " + sUrl);
		m_url->SetValue(sUrl);
	}
    m_browser->LoadURL( m_url->GetValue() );
    m_browser->SetFocus();
    UpdateState();
```

**Comment by imReker** [comment]
URL can't be changed in MiniBlink, at least in the version I used.
Maybe try the latest binary from [official](https://github.com/weolar/miniblink49/releases) to test again? 

EDIT:
Asked author of MiniBlink, the request without schema maybe considered as local file:// and OnLoadURLBegin only handle network request.

**Comment by Natulux** [comment]
Actually I dont find myself able to download binaries from MiniBlink, mostly because I cant read it. I clicked the link with "SDK" but the download page wants me to log in or something and the popup can't be translated by google.. :-/

**Comment by Natulux** [comment]
> EDIT:
> Asked author of MiniBlink, the request without schema maybe considered as local file:// and OnLoadURLBegin only handle network request.

And set a standard scheme or overwriting the scheme is not possible?

**Comment by imReker** [comment]
Try this. I download the latest SDK.

[miniblink-181214.zip](https://github.com/imReker/wxWebViewBlink/files/2694097/miniblink-181214.zip)

> And set a standard scheme or overwriting the scheme is not possible?

Use RegExp to replace every `src` in html?
Or maybe `wkeNetChangeRequestUrl` to change the URL?

**Comment by Natulux** [comment]
Thank you very much. :-)
Is it enough to just exchange the node.dll?

> Use RegExp to replace every src in html?

It would not only be src, but also href and linking from css and js. But yes, if nothing else works, that might be my last chance to tackle this.

EDIT:
By just replacing the node.dll, I get a bunch of "wke api not found".

**Comment by imReker** [comment]
Use the wke.h in SDK and just test.  My branch of MiniBlink added a lot APIs to fit wxWebView's requirement

**Comment by Natulux** [comment]
When using the new node.dll and the new wke.h instead of wkedefine.h

wkeHasSelection
wkeGetSelectedTextW
wkeGetSelectedText
wkeGetSelectedSourceW
wkeGetSelectedSource

API's are missing. But the app starts regardless.
Unfortunately the behavior in my special case is exactly the same.

**Comment by imReker** [comment]
I'm afraid you can only do regexp replace in your HTML code currently or modify MiniBlink core :(

**Comment by Natulux** [comment]
Sorry for bothering again.
After regEx proved impractical, I made an ugly hack where I intercept 
`wxWebViewMiniBlink::OnNavigation`
(which is the first backend function to be called) and if the URL does not begin with "memory:" I run:
`if (!sUrl.StartsWith("memory:"))`
`{`
`  wkeLoadW(webView, "memory:" + sUrl);`
`  return false;`
`}`

Do you see a problem if the OnNavigation function is repeatedly returning false?

**Comment by imReker** [comment]
The code seems OK, any problem?

**Comment by Natulux** [comment]
Not so far, I just dont think I have gained the nessecary overview to be sure by just try and error on my pc.
Thank you :-)

**Comment by imReker** [comment]
BTW, wxWebViewMiniBlink::OnNavigation will not be triggered when loading js, image or css, it only handle url navigation

**Comment by Natulux** [comment]
I just learned that myself :-/
Index and other html files are shown and navigated just fine, but the content isnt loaded properly. But the handling of any content (like loading css, js or resources) happens in the core of MiniBlink,  isnt it? At least I found no prior event handling function for the loading of kind of resources which I could intercept.

I tried to get a hold on MiniBlink core, but apart from it beeing commented in chinese, my vs2015 compiler stoped after 1 hour of compiling due to heap overflow. ;-)

**Comment by imReker** [comment]
Try wkeNetChangeRequestUrl in wkeOnLoadUrlBegin callback.
See example of wkeOnLoadUrlBegin at https://weolar.github.io/miniblink/doc-main.html
Note that wkeNetChangeRequestUrl only exist in the [latest version](https://github.com/weolar/miniblink49/commit/3551f42b3a6897a16c21b2ea59b95ab3ed635995#diff-29900a0589e792a7432aeee7cbb12cf3) of MiniBlink and there is no precompiled binary currently, so you should compile it by yourself.
To solve heap overflow, decrease CPU core in VS C++ compile settings, I use 2 core on my pc (16GB memory), reboot before compile, and do not do any other thing during compile. You may need 4 hours.

**Comment by Natulux** [comment]
Thanks for the warning. I might want to use some better hardware then. ;-)
I will try as you say and see, if it works. With my attempt of compiling MiniBlink, I just used the default settings and it chose to compile "v8_5_7", even though there is "v8_6_7" already in the commit.
So I compile "v8_6_7", right?

**Comment by imReker** [comment]
Compile project miniblink with Release_vc6 which use v_8_4_8 (You may need build other libs first)

**Comment by Natulux** [comment]
Okey, I could get a server for the compilation and after about 3 hours I got several errors. First I needed to comment:

(WebCookieJarCurlImpl.cpp, line 299)
`// const char* cookieJarFileName = net::WebURLLoaderManager::sharedInstance()->getCookieJarFileName();`
getCookieJarFileName() was commented out everywhere else except here. Probably deprecated.

I have 93 errors more, some of them curious to me, like:
Event.cpp, line 36 +
`std::string Event::returnValueGet() {`
`    return m_returnValue;`
`}`
`void Event::returnValueSet(std::string json) {`
`    m_returnValue = json;`

`    if (m_callback)`
`        (*m_callback)(json);`
`}`
Error msg: 

> C2039	"returnValueGet": Is no element of "mate::Event"
> Project: Electron


Neither those two functions, nor the member variables they are using, are declared. I feel like I mixed incompatible versions here, but I did as you said:
- Release_v6 (win32)
- compiled v8-4-8 first, worked fine
- tried to compile miniblink next <- errors
I restarted the building process several times to make sure, there was no more librarie he could build as a dependency.

If I try to compile the project "miniblink" alone, I just get one more error:

> LNK2001	missing symbol ""public: static enum blink::CodePath __cdecl blink::Character::fixComplexPathToSimplePath(unsigned short *,unsigned int)" (?fixComplexPathToSimplePath@Character@blink@@SA?AW4CodePath@2@PAGI@Z)".	miniblink	D:\MiniBlink\build\miniblink\LayoutText.obj

Oh, btw: I got a Node.dll, but it is only 2.6MB in size.

Also I couldn't find "wkeNetChangeRequestUrl". Did you talk about the original MiniBlink or your branch? I compiled the branch which you linked to from GitHub.

Do I need some common libraries or dependencies? This is a fresh Windows Server 2016, so there might be missing something.

Thanks for your help. I appreciate it!

**Comment by imReker** [comment]
Use the lastest Master branch to compile. wkeNetChangeRequestUrl was added in last week.
I noticed that the description of [latest commit](https://github.com/weolar/miniblink49/commit/545820c485beb9afa8c61dc373aa167eb28dc053) is 'Add missing files and codes'

And, Electron and other exe projects are just (deprecated) examples, so build miniblink dll and libs only. 

BTW,the Master is not always build-able. You can comment out some error code or fix minor problem by yourself.

**Comment by Natulux** [comment]
ok, I was using imReker/miniBlink49 master. Now I checked out weolar/miniBlink49 master.
I was able to compile it without changing anything this time. The node.dll is 22MB in size so I guess, in comparison to the old one with 15MB in size, this version contains a whole lot of new code.

Within my test enviroment, the new node.dll works just as good as the old one, even though I get some debug messages on start up:
"wke api not found" from wkeHasSelection
"wke api not found" from wkeGetSelectedTextW
"wke api not found" from wkeGetSelectedText
"wke api not found" from wkeGetSelectedSourceW
"wke api not found" from wkeGetSelectedSource

Apart from that, I have no further errors.

I found the declaration:
`void wkeNetChangeRequestUrl(wkeNetJob jobPtr, const char* url)`
But I am unsure how to use this function. Is this an event function, which is called when an url request is detected (doesn't look like an event function)?
Or is it rather a function which I can use to change the url when I see fit? In that case I can conveniently change the url in 
`wxWebViewMiniBlink::OnNavigation`
if I like, but I still don't get the page internal calls for sources, right?

If I am right, I still need to tell MiniBlink, that it is supposed to use the memory file system, even though it doesn't read the scheme word "memory:".

[Edit]: I again forgot to update the "wkedefine.h". Now I have done that and I need to solve 5 issues:
![image](https://user-images.githubusercontent.com/12150971/51837809-e3312180-2304-11e9-81bd-66396935bdf3.png)

**Comment by imReker** [comment]
Try wkeNetChangeRequestUrl in wkeOnLoadUrlBegin callback.
See example of wkeOnLoadUrlBegin at https://weolar.github.io/miniblink/doc-main.html

**Comment by Natulux** [comment]
I solved the errors by merging your wkedefine with the wkedefine from the new node.dll project.
But still, I can't get this to work.
Just like before, if I don't give it the appropriate scheme (memory:) then I will never come through "wkeOnLoadUrlbegin".

And in the example of wkeOnLoadUrlBegin, there is a hook for one special url which then is replaced.
But I would need a hook for several hundred links and sources. And I don't even know their names yet. For example: The resources from a Quasar.js created dialog are named with their hash value. It changes each time I build it.
Or did I missunderstand it?

I need a function, where I even come through for loading resources or I need to be able to set a default file system handler (or to keep a given file system handler scheme, once it is given until overwritten by another).

**Comment by Natulux** [comment]
Can you perhaps hint me in the right direction, in which class or function miniblink decides, if it has an appropriate file system handler or if the url is invalid, in which case it returns about:blank?

**Comment by imReker** [comment]
I don't know exactly, I didn't look deep into MiniBlink yet.

Since you can debug MiniBlink now, you can try reverse track by setting up a break point in wkeOnLoadUrlBegin's callback.

Or follow the wkeOnLoadUrlBegin code path:
`wkeOnLoadUrlBegin` ->`CWebView::onLoadUrlBegin` -> `CWebViewHandler::loadUrlBeginCallback` -> `WebURLLoaderManager`/`WebURLLoaderManagerMainTask`

Or maybe simply force `KURL::isLocalFile` return false?

**Comment by Natulux** [comment]
I had an eye on KURL myself. ;-)
Thank you for your help and suggestions!

**Comment by imReker** [comment]
I discussed with the author of MiniBlink, 
You can look at https://github.com/imReker/wxWebViewBlink/blob/master/WebViewMiniBlink.cpp#L698, 
try use wxWebView::SetPage with baseurl `memory:xxxx` to load your Index.html and see if the resources can be loaded correctly.

**Comment by Natulux** [comment]
I didn't notice your answer right away, sorry.
weolar told me, that you convinced him to add scheme functionality to the backend. 
Thanks for that. :-)

At the moment I am waiting for him to release a new version with this feature. I would like to try as you suggested, but I do not have the original node.dll anymore and compiling always takes a lot of time and ressources.

**Comment by imReker** [comment]
You don't have to compile original MiniBlink, download it from the [Release Page](https://github.com/weolar/miniblink49/releases) directly.

**Comment by Natulux** [comment]
I am not sure if I covered this correctly.
Normally I initialize my embedded browser with:
`m_browser->LoadURL("memory:index.html");`

Now I replaced it with:
`wxFSInputStream page("memory:index.html");
m_browser->SetPage(page, "memory:index.html");`

It works, but it doesn't take the "memory:" for further loading purposes. I can see the html from the index.html, but the parts loaded with java script (which is everything with Quasar CLI project) can't find their memory ressources.
